api/src/main/java/bio/terra/tanagra/service/query/EntityCounts.java

```diff
@@ -0,0 +1,85 @@
+package bio.terra.tanagra.service.query;
+
+import bio.terra.tanagra.service.search.Attribute;
+import bio.terra.tanagra.service.search.EntityVariable;
+import bio.terra.tanagra.service.search.Filter;
+import com.google.auto.value.AutoValue;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * A query for counts of entity instances. Contains the number of entity instances in each group,
+ * and a list of the attribute values that define each group.
+ *
+ * <p>e.g. {20, [Male, American]}, {35, [Female, Japanese]}
+ */
+@AutoValue
+public abstract class EntityCounts {
+  /** The primary entity and variable that the counts are being created for. */
+  public abstract EntityVariable primaryEntity();
+
+  /**
+   * Any additional attributes to return as columns of the dataset. These must be attributes of the
```

Comment:
Hmm.. how does this work? Wouldn't it require that we either group by this additional attribute, OR we are already grouping by an attribute that maps 1:1 with this selected attribute?

Reply:
Yes, you can only group by an attribute that maps 1:1 with this selected attribute. My intent here was specifically to support returning e.g. gender name when grouping by gender concept id, so that the UI can use display names.

Now I'm thinking about it, I might be able to add validation for this -- make sure all the additional selected attributes are mapped to values foreign-keyed to one of the group by attributes. I'm going to investigate this some more.

Reply:
OK, this looks like a pretty involved validation change. I was hoping to check the `AttributeMappings` specified in the underlay, to validate that each `additionalSelectedAttribute` maps to a `LookupColumn` that uses a key identical to the column that one of the `groupByAttributes` is mapped to. However, the `AttributeMappings` are stored separately from the `Entity` object, so I don't have access to it in this `EntityCounts` class. Another (I think less ideal) option is to validate this during the SQL query construction. However, the way the visitor pattern is currently setup, when resolving a `Selection` expression, which includes `GROUP BY` and `SELECT` fields, we don't have references to other fields or their mappings.

I could change the code to make either of these approaches possible -- the relevant information is in the underlay, I just need to plumb it through to where we want to do the validation. However, I think a better approach is to add the idea of an attribute that is a bundle of fields (e.g. Gender attribute = `gender_concept_id` field and FK'ed `gender_concept_name`). When someone wants to select the Gender attribute, we always include both fields. If we implement this, then we wouldn't need the `additionalSelectedAttributes` API field, and this validation would be moot. It would also help when querying entity instances -- the UI would just request the Gender attribute, instead of having to remember to request both the `gender_concept_id` and the `gender_concept_name`. What do you think?

Reply:
I think the bundle approach is the way to go but we should probably approach that as part of the overall underlay work.